File: cpp_opencv/spaces/spaces.cpp
#include <opencv2/opencv.hpp>
#include <stdio.h>
#include <string>

using namespace cv;

std::string photo = "../../photos/dog2.jpg";

int main(){
    Mat image = imread(photo);
    resize(image, image, cv::Size(), 0.2, 0.2);
    imshow("dog", image);

    Mat gray;
    cvtColor(image, gray, COLOR_BGR2GRAY);
    imshow("gray", gray);

    Mat hsv; 
    cvtColor(image, hsv, COLOR_BGR2HSV);
    imshow("hsv", hsv);

    Mat lab;
    cvtColor(image, lab, COLOR_BGR2Lab);
    imshow("lab", lab);

    Mat rgb;
    cvtColor(image, rgb, COLOR_BGR2RGB);
    imshow("rgb", rgb);

    // inverses also work (only no grayscale to hsc)
    Mat hsv_bgr;
    cvtColor(hsv, hsv_bgr, COLOR_HSV2BGR);
    imshow("hsv_bgr", hsv_bgr);

    Mat lab_bgr; 
    cvtColor(lab, lab_bgr, COLOR_Lab2BGR);
    imshow("lab2bgr", lab_bgr);


    waitKey(0);
    return 0;
}

File: py_opencv/spaces.py
import cv2 as cv 
# import matplotlib.pyplot as plt
import numpy as np 



img = cv.imread("photos/dog2.jpg")
cv.imshow("dog", img)

# plt.imshow(img)
# plt.show()

""" gray scale """
gray = cv.cvtColor(img, cv.COLOR_BGR2GRAY)
cv.imshow('gray', gray)

""" BGR to HSV hue saturation value"""
hsv = cv.cvtColor(img, cv.COLOR_BGR2HSV)
cv.imshow ('hsv', hsv)

""" BGR to LAB """
lab = cv.cvtColor(img, cv.COLOR_BGR2LAB)
cv.imshow('lab', lab)

""" BGR to RGB """ # Open cv default is bgr, matplot lib is rgb 
rgb = cv.cvtColor(img, cv.COLOR_BGR2RGB)
cv.imshow('rgb', rgb)
# plt.imshow(rgb)
# plt.show()

# can do all the inverses, only cannot convert grayscale to hsv directly 
""" HSV to BGR """
hsv_bgr = cv.cvtColor(hsv, cv.COLOR_HSV2BGR)
cv.imshow('hsv2bgr', hsv_bgr)

""" LAB to BGR """
lab_bgr = cv.cvtColor(lab, cv.COLOR_LAB2BGR)
cv.imshow('lab_bgr', lab_bgr)

cv.waitKey(0)